crates/pathfinder/src/state/sync/pending.rs

```diff
@@ -28,96 +31,147 @@ pub async fn poll_pending(
+                        tracing::trace!(hash=%block.block_hash, "Found current head from pending mode");
+                    }
+                    MaybePendingBlock::Block(block) => {
+                        tracing::trace!(hash=%block.block_hash, "Found full block, exiting pending mode.");
+                        return Ok((Some(block), None));
+                    }
+                    MaybePendingBlock::Pending(pending) if pending.parent_hash != head.0 => {
+                        tracing::trace!(
+                            pending=%pending.parent_hash, head=%head.0,
+                            "Pending block's parent hash does not match head, exiting pending mode"
+                        );
+                        return Ok((None, None));
+                    }
+                    MaybePendingBlock::Pending(pending) => {
+                        let replace = match &prev_block {
+                            None => true,
+                            Some(prev) if pending.transactions.len() > prev.transactions.len() => true,
+                            _ => false,
+                        };
+
+                        if replace {
+                            prev_block = Some(Arc::new(pending));
+                            changed = true;
+                            tracing::trace!("Pending block data changed");
+                        } else {
+                            tracing::trace!("No change in pending block data");
+                        }
+                    },
+                }
+
+                let gateway_copy = sequencer.clone();
+                block_task = tokio::spawn(async move {
+                    tokio::time::sleep_until(Instant::now() + poll_interval).await;
+                    gateway_copy
+                        .block(BlockId::Pending)
+                        .await
+                        .context("Downloading pending block")
+                });
+            },
+
+            state_update = &mut update_task => {
+                let  state_update = state_update.context("Joining state update download task")??;
+
+                if state_update.block_hash != pathfinder_common::BlockHash::ZERO {
+                    tracing::trace!("Found full state update, exiting pending mode.");
+                    return Ok((None, Some(state_update)));
+                } else if state_update.parent_state_commitment != head.1 {
+                    tracing::trace!(pending=%state_update.parent_state_commitment, head=%head.1, "Pending state update's old root does not match head, exiting pending mode.");
+                    return Ok((None, None));
+                }
+
+                let replace = match &prev_state_update {
+                    None => true,
+                    Some(prev) if state_update.change_count() > prev.change_count() => true,
+                    _ => false,
+                };
```

Comment:
Or even this:

```suggestion
                let replace = prev_state_update.as_ref()
                    .map(|prev| state_update.change_count() > prev.change_count())
                    .unwrap_or(true);
```

Reply:
Thanks, was looking for something like this. My brain was not firing..